app/rdf/query-cube-preview.ts

```diff
@@ -118,130 +126,108 @@ export const getCubePreview = async (
         }}
       }
     }`,
-    { operation: "postUrlencoded" }
-  );
+      { operation: "postUrlencoded" }
+    )
+    .then((qs) => {
+      return qs.map(({ subject, predicate, object, ...rest }) => {
+        return { ...rest, s: subject, p: predicate, o: object };
+      });
+    });
 
-  if (quads.length === 0) {
+  if (qs.length === 0) {
     throw new Error(`No cube found for ${iri}!`);
   }
 
-  const dimensionsQuads = quads.filter(
-    (quad) =>
-      quad.predicate.equals(ns.sh.path) &&
-      !quad.object.equals(ns.cube.observedBy) &&
-      !quad.object.equals(ns.rdf.type)
+  const qsDims = qs.filter(
+    ({ p, o }) =>
+      p.equals(ns.sh.path) &&
+      !o.equals(ns.cube.observedBy) &&
+      !o.equals(ns.rdf.type)
   );
   const dimensions: Dimension[] = [];
   const measures: Measure[] = [];
-  uniqBy(dimensionsQuads, (quad) => quad.object.value).map((quad) => {
-    const dimensionQuads = quads.filter((q) => q.subject.equals(quad.subject));
-    const dimensionIri = quad.object.value;
-    const labelQuad = dimensionQuads.find((q) =>
-      q.predicate.equals(ns.schema.name)
-    );
-    const descriptionQuad = dimensionQuads.find((q) =>
-      q.predicate.equals(ns.schema.description)
-    );
-    const orderQuad = dimensionQuads.find((q) =>
-      q.predicate.equals(ns.sh.order)
-    );
-    const typeQuads = dimensionQuads.filter((q) =>
-      q.predicate.equals(ns.rdf.type)
-    );
-    const scaleTypeQuad = dimensionQuads.find((q) =>
-      q.predicate.equals(ns.qudt.scaleType)
-    );
-    const scaleType = getScaleType(scaleTypeQuad?.object);
-    const dataTypeQuad = dimensionQuads.find((q) =>
-      q.predicate.equals(ns.sh.datatype)
-    );
-    const unitQuad = dimensionQuads.find((q) =>
-      q.predicate.equals(ns.qudt.unit)
-    );
-    const unitLabelQuad = quads.find((q) => q.subject.equals(unitQuad?.object));
-    const dataKindQuad = dimensionQuads.find((q) =>
-      q.predicate.equals(ns.cube("meta/dataKind"))
-    );
-    const dataKindTypeQuad = quads.find(
-      (q) =>
-        q.subject.equals(dataKindQuad?.object) &&
-        q.predicate.equals(ns.rdf.type)
+  uniqBy(qsDims, ({ o }) => o.value).map(({ s, o }) => {
+    const dimIri = o.value;
+    const qsDim = qs.filter((q) => q.s.equals(s));
+    const qLabel = qsDim.find((q) => q.p.equals(ns.schema.name));
+    const qDesc = qsDim.find((q) => q.p.equals(ns.schema.description));
+    const qOrder = qsDim.find((q) => q.p.equals(ns.sh.order));
+    const qsType = qsDim.filter((q) => q.p.equals(ns.rdf.type));
+    const qScaleType = qsDim.find((q) => q.p.equals(ns.qudt.scaleType));
+    const scaleType = getScaleType(qScaleType?.o);
+    const qDataType = qsDim.find((q) => q.p.equals(ns.sh.datatype));
+    const qUnit = qsDim.find((q) => q.p.equals(ns.qudt.unit));
+    const qUnitLabel = qs.find((q) => q.s.equals(qUnit?.o));
+    const qDataKind = qsDim.find((q) => q.p.equals(ns.cube("meta/dataKind")));
+    const qDataKindType = qs.find(
```

Comment:
Can't we use `const qlabel = qsDim.out(ns.schema.name)` ? Here, we iterate over the predicates of the dimension for each dimension for each dimesion predicate.

Reply:
We can't use `.out` here, as it's a method of a [`Dataset`](https://github.com/zazuko/clownface/blob/1f1a8a6991f9a506942eeec1a4afc113a43cc2e4/lib/Context.js#L30-L40), not `Quad` that we found above. As in this query we fetch several not related things (e.g. data related to cube, dimensions, observations), we would need to create several instances of `Dataset` class to be able to use `.out` – I tried to use this approach at first, but I had problems to make it work correctly (and it [seems to be doing the same thing](https://github.com/zazuko/clownface/blob/1f1a8a6991f9a506942eeec1a4afc113a43cc2e4/lib/Context.js#L121-L148), just wrapped). I could try to rewrite the logic, but I'm not 100% sure it would be more performant 🤔

Reply:
OK, I did not know that in/out were iterating over all the current context predicates, thanks for pointing this out.